FirebaseAppDistribution/Sources/FIRAppDistributionMachO.m

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <CommonCrypto/CommonHMAC.h>
+#include <mach-o/arch.h>
+#import <mach-o/fat.h>
+#import <mach-o/loader.h>
+#import "FIRAppDistributionMachO+Private.h"
+#import "FIRAppDistributionMachOSlice+Private.h"
+
+@implementation FIRAppDistributionMachO
+
+- (instancetype)initWithPath:(NSString*)path {
+  self = [super init];
+
+  if (self) {
+    _file = [NSFileHandle fileHandleForReadingAtPath:path];
+    [_file seekToFileOffset:0];
+    _slices = [NSMutableArray new];
+    [self extractSlices];
+  }
+
+  return self;
+}
+
+- (void)extractSlices {
+  uint32_t magicValue;
+
+  struct fat_header fheader;
+  NSData* data = [_file readDataOfLength:sizeof(fheader)];
+  [data getBytes:&fheader length:sizeof(fheader)];
+
+  magicValue = CFSwapInt32BigToHost(fheader.magic);
+
+  if (magicValue == FAT_MAGIC) {
+    uint32_t archCount = CFSwapInt32BigToHost(fheader.nfat_arch);
+    NSUInteger archOffsets[archCount];
+
+    // Gather the offsets
+    for (uint32_t i = 0; i < archCount; i++) {
```

Comment:
for testability, we could pull this out into another function so we can feed back however many offsets we want to the next chunk - it may not be worth it but something to think about

Reply:
Similar to above, I'm hoping we don't have to test the "how many architectures did we find and what are there offsets", and more just "we put in binary X and it produced hash Y"